lib/features/mailbox/data/repository/mailbox_repository_impl.dart

```diff
@@ -3,89 +3,140 @@ import 'package:jmap_dart_client/jmap/account_id.dart';
+  }) async {
+    if (mailboxUpdated != null && mailboxUpdated.isNotEmpty) {
+      final newMailboxUpdated = mailboxUpdated.map((mailboxUpdated) {
+        if (updatedProperties == null) {
+          return mailboxUpdated;
+        } else {
+          final mailboxOld = mailboxCacheList?.findMailbox(mailboxUpdated.id);
+          if (mailboxOld != null) {
+            return mailboxOld.combineMailbox(mailboxUpdated, updatedProperties);
+          } else {
+            return mailboxUpdated;
+          }
+        }
+      }).toList();
+
+      return newMailboxUpdated;
+    }
+    return mailboxUpdated;
   }
 
   @override
   Stream<MailboxResponse> refresh(AccountId accountId, State currentState) async* {
-
-    final newChangesResponse = await Future.wait(
-      [
-        mapDataSource[DataSourceType.local]!.getAllMailboxCache(),
-        mapDataSource[DataSourceType.network]!.getChanges(accountId, currentState)
-      ],
-      eagerError: true
-    ).then((List response) async {
-      final cacheResponse = response.first as MailboxCacheResponse;
-      final mailboxList = cacheResponse.mailboxCaches != null
-          ? cacheResponse.mailboxCaches!.toMailboxList()
-          : <Mailbox>[];
-
-      final changeResponse = response.last as MailboxChangeResponse;
-
-      return await mapDataSource[DataSourceType.local]!.combineMailboxCache(changeResponse, mailboxList);
+    final localMailboxResponse = await Future.wait([
+      mapDataSource[DataSourceType.local]!.getAllMailboxCache(),
+      stateDataSource.getState(StateType.mailbox)
+    ]).then((List response) {
+      return MailboxResponse(mailboxes: response.first, state: response.last);
     });
 
-    await mapDataSource[DataSourceType.local]!.asyncUpdateCache(newChangesResponse);
+    bool hasMoreChanges = true;
+    State? sinceState = localMailboxResponse.state!;
 
-    final newMailboxCacheResponse = await mapDataSource[DataSourceType.local]!.getAllMailboxCache();
+    while(hasMoreChanges && sinceState != null) {
+      final changesResponse = await mapDataSource[DataSourceType.network]!.getChanges(accountId, sinceState);
 
-    final newMailboxList = newMailboxCacheResponse.mailboxCaches != null
-        ? newMailboxCacheResponse.mailboxCaches!.toMailboxList()
-        : <Mailbox>[];
-    final newState = newMailboxCacheResponse.oldState?.state;
+      hasMoreChanges = changesResponse.hasMoreChanges;
+      sinceState = changesResponse.newStateChanges;
+
+      final newMailboxUpdated = await _combineMailboxCache(
+          mailboxUpdated: changesResponse.updated,
+          updatedProperties: changesResponse.updatedProperties,
+          mailboxCacheList: localMailboxResponse.mailboxes!);
+
+      await Future.wait([
+        mapDataSource[DataSourceType.local]!.update(
+            updated: newMailboxUpdated,
+            created: changesResponse.created,
+            destroyed: changesResponse.destroyed),
+        if (changesResponse.newStateMailbox != null)
```

Comment:
if update failed, IMO, we should not save the state. So I think we need to wait update mailbox completed then update state. WDYT?

Reply:
I don't think that's necessary. Because when we get all mailbox mailboxes, we have checked the state is nulll or not.

Reply:
I mean the case:
1. you refresh with `/changes`
2. while you update the cache, something wrong, cache is not update correctly (some mailbox is not cache)
3. if you update `newState` on that case then in the next time of `/changes` you should miss some data is not updated in (2)

Reply:
I understand you. But it's no problem at all. On the other hand, hive database does not return bool when inserting an object.